src/Compilers/Core/Portable/CodeGen/PrivateImplementationDetails.cs

```diff
@@ -292,8 +319,131 @@ internal Cci.IFieldReference CreateDataField(ImmutableArray<byte> data, ushort a
+
+            if (method is null)
+            {
+                TryAddSynthesizedMethod(synthesizeBytesToStringHelper(this, compilation, syntaxNode, diagnostics));
+
+                method = GetMethod(SynthesizedBytesToStringFunctionName);
+            }
+
+            Debug.Assert(method is not null);
+            return method;
+
+            static Cci.IMethodDefinition synthesizeBytesToStringHelper(
+                PrivateImplementationDetails @this,
+                Compilation compilation,
+                SyntaxNode syntaxNode,
+                DiagnosticBag diagnostics)
+            {
+                if (!tryGetWellKnownTypeMember(compilation, syntaxNode, diagnostics, WellKnownMember.System_Text_Encoding__get_UTF8, out var encodingUtf8) |
+                    !tryGetWellKnownTypeMember(compilation, syntaxNode, diagnostics, WellKnownMember.System_Text_Encoding__GetString, out var encodingGetString))
+                {
+                    return new DefaultMethodDef(
+                        name: SynthesizedBytesToStringFunctionName,
+                        visibility: Cci.TypeMemberVisibility.Private,
+                        containingType: @this,
+                        maxStack: 8,
+                        il: []);
+                }
+
+                Debug.Assert(encodingUtf8 is not null && encodingGetString is not null);
+
+                var ilBuilder = new ILBuilder(
+                    (ITokenDeferral)@this._moduleBuilder,
+                    new LocalSlotManager(slotAllocator: null),
+                    OptimizationLevel.Release,
+                    areLocalsZeroed: false);
+
+                // Call `Encoding.get_UTF8()`.
+                ilBuilder.EmitOpCode(ILOpCode.Call, 1);
+                ilBuilder.EmitToken(encodingUtf8, null, diagnostics);
+
+                // Push the `byte*`.
+                ilBuilder.EmitOpCode(ILOpCode.Ldarg_0);
+
+                // Push the byte size.
+                ilBuilder.EmitOpCode(ILOpCode.Ldarg_1);
+
+                // Call `Encoding.GetString(byte*, int)`.
+                ilBuilder.EmitOpCode(ILOpCode.Callvirt, -2);
+                ilBuilder.EmitToken(encodingGetString, null, diagnostics);
+
+                // Return.
+                ilBuilder.EmitRet(isVoid: false);
+                ilBuilder.Realize();
+
+                return new BytesToStringHelper(
+                    containingType: @this,
+                    encodingGetString: encodingGetString,
+                    maxStack: ilBuilder.MaxStack,
+                    il: ilBuilder.RealizedIL);
+            }
+
+            static bool tryGetWellKnownTypeMember(
+                Compilation compilation,
+                SyntaxNode syntaxNode,
+                DiagnosticBag diagnostics,
+                WellKnownMember member,
+                [NotNullWhen(returnValue: true)] out Cci.IMethodReference? result)
+            {
+                if (compilation.CommonGetWellKnownTypeMember(member) is { } symbol)
```

Comment:
>if (compilation.CommonGetWellKnownTypeMember(member) is { } symbol) [](http://example.com/codeflow?start=16&length=67)

In general, I think there is no guarantee that the symbol returned by ```CommonGetWellKnownTypeMember``` is good. We should find a way to do what ```Binder.GetWellKnownTypeMember``` is doing.